### ISSUE TITLE
Add multi-package support to rez-pip

### ISSUE THREAD

**Issue by instinct-vfx** [nerdvegas/rez#682]
Extend rez-pip to support multi-package installations using both requirements.txt and adding multiple packages to the cmdline. e.g.

`rez-pip -i -r arrow Click`

I am unsure how to best handle requirements.txt files. The original argument is -r which rez-pip interprets as "release" already. Opinions?

**Comment by nerdvegas** [comment]
I'm cool with adding multiple package support to CLI, but is there a
specific need for parsing requirements from a requirements.txt?

A

**Comment by instinct-vfx** [comment]
The main reason for me is that we have quite a few packages that are python packages, come with python dependencies (and sometimes quite a lot of dependencies) but are not installed through rez-pip. An example would be packages that need to be bootstrapped into a DCC. These would not be installed through rez-pip because the package.py needs to do things that pip does not support.

For these cases installing dependencies can be quite a hassle. We oftentimes still build these packages with full pip support so there already is a requirements.txt file. Installing dependencies like this would be convenient.

Another option would be to integrate this into our rezbuild setups. Having a shared function that can reads the requirements.txt and calls rez-pip on each of those. My gut feeling is that this would be a weird rez-ception at build-time, but i may be off. What do you think?

**Comment by nerdvegas** [comment]
So first, that use case is reasonable so I think that's reason enough for
requirements.txt support in rez-pip. The work needed for that has already
been done (see https://github.com/nerdvegas/rez/pull/667, which will be
merged soon). I may even make it part of the same PR.

In terms of a rez package that uses pip to install its dependencies. Ideas
similar to this have been raised in the past, and recently as well. In fact
you can take the idea further, and add pluggable support for setup.py-
based packages, that rez would then be able to ingest and build natively.
My thoughts on this at the moment are that it's an interesting place to go,
and I think it makes sense - ie, the general case where a non-package.py
file can be run through a step that converts it into a rez package. However
bear in mind that rez was not designed specifically for managing python
packages, and so any work in this area needs to be done with that in mind
(hence framing this as a general, plugin-driven approach).

One thing to mention also is that rez-build does NOT build a package's
dependencies also, and should not. That is not the job of the build. That's
not to say we don't want that functionality however - in fact, the main
part of my rez talk here at siggraph tomorrow is all about how we implement
a `rez-install` tool that is capable of that kind of thing (ie, installing
rez packages and their requirements from public rez package repositories).

Cheers
A

**Comment by lambdaclan** [comment]
Hello Allan, just want to make sure before I start working on it but is your pip improvements PR #667  going to add support for multi package install and installing from a file ala requirements.txt 

```
rez-pip -i arrow click django
rez-pip -i ../project/requirements.txt # (same format as usual)
```

or do you mean the necessary changes to support said actions will be complete but the implementation still pending?

> I think that's reason enough for requirements.txt support in rez-pip. The work needed for that has already been done (see #667, which will be merged soon). I may even make it part of the same PR.

If it is the latter, once your PR is merged I will be happy to look into the implementation so I will wait until I hear from you.

**Comment by nerdvegas** [comment]
Sorry getting back late, missed this thread.

Yes I think we should do this in separate PR once #667 is merged, cheers!

**Comment by lambdaclan** [comment]
Sounds good, I will start working on a new PR to add said functionality once #667 is merged then 👍

**Comment by nebukadhezer** [comment]
@lambdaclan have you been looking into this recently ?
